[LoopVectorize] LLVM fails to vectorise loops with multi-bool varables

dtemirbulatov opened the pull request:
This patch allows to consider compare instructions in the loop with multiple use inside the loop and outside.

This change allows to vectorise this loop:
int foo(float* a, int n) {
  _Bool any = 0;
  _Bool all = 1;
  for (int i = 0; i < n; i++) {
    if (a[i] < 0.0f) {
      any = 1;
    } else {
      all = 0;
    }
  }
  return all ? 1 : any ? 2 : 3;
}

Commit:
[LoopVectorize] LLVM fails to vectorise loops with multiple bool variables

This patch allows to consider compare instructions in the loop with multiple
use inside the loop and outside, if we can prove that compare instruction user
is a recurrent reduction or used in branching or outside the loop then it is
safe to consider to vectorise.

This change allows to vectorise this loop:
int foo(float* a, int n) {
  _Bool any = 0;
  _Bool all = 1;
  for (int i = 0; i < n; i++) {
    if (a[i] < 0.0f) {
      any = 1;
    } else {
      all = 0;
    }
  }
  return all ? 1 : any ? 2 : 3;
}

SamTebbs33 commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -635,14 +639,59 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
       return InstDesc(Select, Prev.getRecKind());
   }
 
+  SelectInst *SI = dyn_cast<SelectInst>(I);
```
I think this change could do with some comments.

dtemirbulatov replied:
Done.

SamTebbs33 commented on llvm/test/Transforms/LoopVectorize/multicmp.ll:
```diff
@@ -0,0 +1,794 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
```
Could you pre-commit this test so we can see the difference your patch makes?

dtemirbulatov replied:
Thanks, Sam. I just added test description or C representations. Is it ok?

SamTebbs33 replied:
The comments are useful but I meant that you could commit the tests to main and rebase this PR on top of that so that we can see the difference this PR makes. Florian made a similar suggestion but with a new PR instead of committing straight to main.

david-arm replied:
Or just take the simplest option of having two commits in this patch. Commit 1 - add tests, Commit 2 - add the code change which shows the test changes.

SamTebbs33 replied:
That's a better idea. I'm used to the fabricator days when you couldn't do that!

dtemirbulatov replied:
Done.

fhahn replied:
> That's a better idea. I'm used to the fabricator days when you couldn't do that!

Splitting the test changes completely off the functional changes has a few advantages compared to adding them as separate commit to this PR:
1. Having a separate PR avoids cluttering the functional review with a large number of comments regarding the tests
2. Github's PR interface shows the changes from all commits in the `Files changed` view
3. As all commits in a PR get squashed together, the commit landing won't just show the diff on the tests caused by the patch, making it more difficult for people looking at the commit later (e.g. when doing post-commit review).

david-arm replied:
The only problem is that you'll probably still end up with lots of comments or requests to change the tests once reviewers see the actual code changes and, for example, realise that a test case is missing for a particular change. Then you still end up changing the tests in the second patch anyway and at that point it's too late to change the first test-only patch because it has already landed.

I realise the official solution to this is to create branches in the main llvm repo in order to do stacked pull requests. That technically is a solution that probably works well for some cases, but given this is a very frequent thing to do it's also more costly and it seems that people aren't deleting their branches afterwards. `git branch -r -l | grep users | wc -l` returns a count of 638.

Not trying to say which way is right or wrong, or what works best for the individual developer, but just pointing out there are pros and cons to each approach.

SamTebbs33 replied:
I agree with David's concern. Regarding your points fhahn, one can see the differences between commits in the "files changed" view and there is a link to the PR in the commit body so people can see the test diff during post-commit review.

Commit:
Resolved remarks.

dtemirbulatov commented:
Ping.

david-arm reviewed (commented):
Thanks for this @dtemirbulatov - it looks like a useful addition! I've not done a full review of the code changes, but I've left some comments so far.

fhahn commented:
Could you split off the new tests in a separate PR so we can only see the test changes here? Also would probably be good to have at least some target specific tests to check cost modeling

Commit:
Resolved remarks.

dtemirbulatov commented on llvm/test/Transforms/LoopVectorize/multicmp.ll:
```diff
@@ -0,0 +1,883 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=1 -force-vector-width=4 -S < %s | FileCheck %s --check-prefix=CHECK
```
Done.

dtemirbulatov commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -270,6 +275,8 @@ class RecurrenceDescriptor {
            cast<IntrinsicInst>(I)->getIntrinsicID() == Intrinsic::fmuladd;
   }
 
+  Instruction *getMultiCmp() const { return MultiCmp; }
```
Done.

dtemirbulatov commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -656,8 +711,10 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+  return InstDesc(I,
+                  isa<ICmpInst>(I->getOperand(0)) ? RecurKind::IAnyOf
+                                                  : RecurKind::FAnyOf,
+                  nullptr, Cmp);
```
Done.

Commit:
Changed multi_user_cmp_max() function to use llvm.smax.i32 intrinsic,
Replaced std::find(Blocks.begin(), Blocks.end(),... to
Loop->contains(Inst->getParent()), added comments.

Commit:
Fix a compile time failure, while testing proposed change.

Commit:
Original verison of llvm/test/Transforms/LoopVectorize/multicmp.ll

dtemirbulatov commented:
Ping.

sdesmalen-arm commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -635,14 +635,57 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+          SelectNonPHIUserInLoop = true;
+      }
+    }
+    Cmp = dyn_cast<CmpInst>(SI->getOperand(0));
```
If the compare used in the `select` has another use which is not part of any reduction, does that mean it's no longer an "any_of" pattern? I would expect `isAnyOfPattern` to still return `true`, but for any further checks to be done somewhere else.

sdesmalen-arm commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -635,14 +635,57 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+        if (!UInst)
+          continue;
+        if (SelectInst *SI1 = dyn_cast<SelectInst>(U)) {
+          if (!llvm::all_of(SI1->users(), [Loop](User *USI) {
```
Having a loop that goes through all users of all users of the Compare seems rather expensive.

sdesmalen-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1045,6 +1060,23 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
   }
 
+  // Make sure that all compare instruction users are recurrent if in loop's BB.
+  if (MultiCmpsRed.size() > 0) {
```
Why do we need similar loops/checks in both LoopVectorizationLegality.cpp and IVDecriptors.cpp?

dtemirbulatov replied:
There might be other kind of reduction and compare instruction with multipe users, I have seen such a case. I think I will add a test for this case.

sdesmalen-arm reviewed (changes requested):
Because of missing negative tests I am not really sure which cases you're trying to prevent from being vectorised. The code itself is a little too incomprehensible to make that purpose clear to me either.

Commit:
Update tests.

dtemirbulatov commented:
We found issue with this enablement. For this example:
```
void foo(int *restrict a, int *restrict j, int n) {
 _Bool any = 0;
 _Bool all = 1;
 for (int i = 0; i < n; i++) {
  _Bool c = a[i] < 0;
  if (c)
   any = 1;
  else
   all = 0;
  j[i] = all ? 1 : any ? 2 : 3;
 }
}
```
In vectorized version of this loop, LV missed to introduce reductions for all and any variables.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/select-multi-cmp.ll:
Perhaps you can move this test to Transforms/LoopVectorize/AArch64/select-costs.ll and delete this file?

dtemirbulatov replied:
Done.

david-arm commented on llvm/test/Transforms/LoopVectorize/multicmp.ll:
I tried running these locally and the tests failed. I think this patch just needs a rebase.

dtemirbulatov replied:
Done.

david-arm reviewed (commented):
I think the patch looks good and the tests are comprehensive! I just had a couple of minor comments, in particular the patch needs rebasing. Thanks!

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/select-costs.ll:
```diff
@@ -47,3 +43,47 @@ for.cond.cleanup.loopexit:                        ; preds = %for.body
+; CHECK-NEXT: LV: Found an estimated cost of 0 for VF 16 For instruction:   %arrayidx = getelementptr inbounds float, ptr %a, i64 %indvars.iv
+; CHECK-NEXT: LV: Found an estimated cost of 4 for VF 16 For instruction:   %load1 = load float, ptr %arrayidx, align 4
+; CHECK-NEXT: LV: Found an estimated cost of 4 for VF 16 For instruction:   %cmp1 = fcmp olt float %load1, 0.000000e+00
+; CHECK-NEXT: LV: Found an estimated cost of 1 for VF 16 For instruction:   %.any.0.off0 = select i1 %cmp1, i1 true, i1 %any.0.off09
+; CHECK-NEXT: LV: Found an estimated cost of 1 for VF 16 For instruction:   %all.0.off0. = select i1 %cmp1, i1 %all.0.off010, i1 false
```
Thanks for moving the test! I think if what you really want to test is the cost of the select instructions, perhaps it's better to write the costs in the same way as `@selects_1`, i.e.

```
; CHECK-NEXT: LV: Found an estimated cost of 1 for VF 8 For instruction:   %.any.0.off0 = select i1 %cmp1, i1 true, i1 %any.0.off09
; CHECK-NEXT: LV: Found an estimated cost of 1 for VF 8 For instruction:   %all.0.off0. = select i1 %cmp1, i1 %all.0.off010, i1 false

; CHECK-NEXT: LV: Found an estimated cost of 1 for VF 16 For instruction:   %.any.0.off0 = select i1 %cmp1, i1 true, i1 %any.0.off09
; CHECK-NEXT: LV: Found an estimated cost of 1 for VF 16 For instruction:   %all.0.off0. = select i1 %cmp1, i1 %all.0.off010, i1 false
```

dtemirbulatov replied:
Done.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/select-costs.ll:
```diff
@@ -47,3 +43,47 @@ for.cond.cleanup.loopexit:                        ; preds = %for.body
+  ret i32 %1
+}
+
+; CHECK-LABEL: define void @selects_1(
```
I see why you had to move these CHECK lines. Perhaps you don't really need them at all? Since you're collecting the debug output from the vectoriser anyway, you can simply CHECK for the chosen VF, i.e.

```
LV: Found an estimated cost of 1 for VF 16 For instruction:   %.any.0.off0 = select i1 %cmp1, i1 true, i1 %any.0.off09
LV: Found an estimated cost of 1 for VF 16 For instruction:   %all.0.off0. = select i1 %cmp1, i1 %all.0.off010, i1 false
...
LV: Selecting VF: 16
```

This way you don't need to check any lines of IR and might simplify the test.

david-arm replied:
I think this comment still stands. You can simplify the test by removing these check lines and simply adding a CHECK line for the selected VF to the existing CHECK lines above. If you do this then you can also simplify the test RUN line to be:

`; RUN: opt < %s -passes=loop-vectorize -debug-only=loop-vectorize -disable-output -S 2>&1 | FileCheck %s`

david-arm commented on llvm/test/Transforms/LoopVectorize/select-cmp.ll:
```diff
@@ -247,6 +247,246 @@ exit:                                     ; preds = %for.body
 }
 
 
+define i32 @select_const_i32_from_icmp_mul_use(ptr nocapture readonly %v1, ptr %v2, i64 %n) {
```
It looks like you're already testing this in the new multicmp.ll test file. Perhaps you can just delete the test?

dtemirbulatov replied:
Done.

david-arm commented on llvm/test/Transforms/LoopVectorize/multicmp.ll:
It might be worth renaming this file to select-cmp-multiuse.ll? That way it's a bit more consistent with the select-cmp.ll naming and shows the difference between them?

david-arm reviewed (commented):
Thanks for the rebase and making the changes! I've got a few more comments, but once they're done I'm happy. :)

fhahn commented:
@dtemirbulatov I think there was an earlier comment about an issue you found while testing this, but it seems to have disappeared? Just double checking if the latest version includes a fix + test for it?

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/select-costs.ll:
```diff
@@ -47,3 +43,38 @@ for.cond.cleanup.loopexit:                        ; preds = %for.body
 }
+
+define i32 @multi_user_cmp(ptr readonly %a, i32 noundef %n) {
+; CHECK: LV: Found an estimated cost of 4 for VF 16 For instruction:   %cmp1 = fcmp olt float %load1, 0.000000e+00
```
Sorry @dtemirbulatov I just have one more comment. I guess it may be my fault you're hitting this issue, but at the moment these CHECK lines are fragile and it's not obvious which function we are actually testing. That's because these CHECK lines could in theory match the CHECK lines in `@selects_1`. In each function the CHECK line should start with:

```
define void @selects_1(ptr nocapture %dst, i32 %A, i32 %B, i32 %C, i32 %N) {
; CHECK-LABEL: LV: Checking a loop in 'selects_1'
... existing CHECKs ...
...
}

define i32 @multi_user_cmp(ptr readonly %a, i32 noundef %n) {
; CHECK-LABEL: LV: Checking a loop in 'multi_user_cmp'
... existing CHECKs ...
...
}

dtemirbulatov replied:
Done.

david-arm reviewed (commented):
Thanks for the updates @dtemirbulatov! I just have a few minor comments left on the cost test file.

Commit:
Further test updates.

david-arm reviewed (approved):
LGTM! Thanks for all the changes @dtemirbulatov.

fhahn reviewed (approved):
LGTM, thanks!

There's one place where we modify the matched compare for any_of reductions (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L9034) which I think relies on the select being the single user. I don't think cases where this triggers are currently supported by this change, but it would probably be good to add an assert there that the modified compare has a single user.

dtemirbulatov commented on llvm/test/Transforms/LoopVectorize/AArch64/select-costs.ll:
```diff
@@ -47,3 +45,32 @@ for.cond.cleanup.loopexit:                        ; preds = %for.body
+
+for.body:
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %for.body ]
+  %all.0.off010 = phi i1 [ true, %entry ], [ %all.0.off0., %for.body ]
```
Done.

dtemirbulatov commented on llvm/test/Transforms/LoopVectorize/AArch64/select-costs.ll:
```diff
@@ -47,3 +45,32 @@ for.cond.cleanup.loopexit:                        ; preds = %for.body
+; CHECK: LV: Found an estimated cost of 1 for VF 16 For instruction:   %all.0.off0. = select i1 %cmp1, i1 %all.0.off010, i1 false
+; CHECK: LV: Selecting VF: 16.
+entry:
+  %wide.trip.count = zext nneg i32 %n to i64
```
Done.

dtemirbulatov commented:
@fhahn I don't think we need this extra assert at (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L9034).

fhahn commented:
> @fhahn I don't think we need this extra assert at (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L9034).

Yep I don't think it is needed in the VPlan-native path/VPInstructionsToVPRecipes
